Greetings! This is going to be an exploration of an eBay car sales dataset. This project will involve the cleaning and analyzing of used car listings to uncover insights. There are no guiding hypotheses; only an appetite to learn and uncover the story behind the data. The dataset is a smaller sample provided by the kaggle dataset (https://www.kaggle.com/orgesleka/used-cars-database/data#autos.csv).

In [218]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


%matplotlib inline

In [219]:
autos = pd.read_csv("autos_2.csv", encoding = "Latin-1")
autos.head()

Unnamed: 0          dateCrawled  \
0           0  2016-03-26 17:47:46   
1           1  2016-04-04 13:38:56   
2           2  2016-03-26 18:57:24   
3           3  2016-03-12 16:58:10   
4           4  2016-04-01 14:38:50   

                                                name  seller offerType  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   Angebot   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   Angebot   
2                         Volkswagen_Golf_1.6_United  privat   Angebot   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   Angebot   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...  privat   Angebot   

    price   abtest vehicleType  yearOfRegistration    gearbox  \
0  $5,000  control         bus                2004    manuell   
1  $8,500  control   limousine                1997  automatik   
2  $8,990     test   limousine                2009    manuell   
3  $4,350  control  kleinwagen                2007  automatik   
4  $1,350     test       kombi                2003    manuell   

          ...            model   odometer monthOfRegistration  fuelType  \
0         ...           andere  150,000km                   3       lpg   
1         ...              7er  150,000km                   6    benzin   
2         ...             golf   70,000km                   7    benzin   
3         ...           fortwo   70,000km                   6    benzin   
4         ...            focus  150,000km                   7    benzin   

        brand notRepairedDamage          dateCreated nrOfPictures  postalCode  \
0     peugeot              nein  2016-03-26 00:00:00            0       79588   
1         bmw              nein  2016-04-04 00:00:00            0       71034   
2  volkswagen              nein  2016-03-26 00:00:00            0       35394   
3       smart              nein  2016-03-12 00:00:00            0       33729   
4        ford              nein  2016-04-01 00:00:00            0       39218   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50  

[5 rows x 21 columns]

In [220]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 21 columns):
Unnamed: 0             50000 non-null int64
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int

In [221]:
autos.isnull().sum()/len(autos)*100

Unnamed: 0              0.000
dateCrawled             0.000
name                    0.000
seller                  0.000
offerType               0.000
price                   0.000
abtest                  0.000
vehicleType            10.190
yearOfRegistration      0.000
gearbox                 5.360
powerPS                 0.000
model                   5.516
odometer                0.000
monthOfRegistration     0.000
fuelType                8.964
brand                   0.000
notRepairedDamage      19.658
dateCreated             0.000
nrOfPictures            0.000
postalCode              0.000
lastSeen                0.000
dtype: float64

The dataset contains 50000 observations (each a listing) and 20 variables (each a feature of the listing). Some of the variables have missing values, but none with more than 20%. There also appears to be variables written in "CamelCase" (e.g. eBay, iPhone), versus the preferred python "snake_case" (e.g. e_bay, i_phone). One good place to start is by chaning the camelcase variable names into snakecase format, with names that are also more descriptive. 

## Tidying up the data

In [222]:
#create dictionary
names_dict = {"dateCrawled" : "date_crawled",
              "offerType" : "offer_type", 
              "vehicleType" : "vehicle_type",
              "yearOfRegistration" : "registration_year",
              "powerPS" : "power_ps",
             "monthOfRegistration" : "registration_month",
              "fuelType" : "fuel_type",
             "notRepairedDamage" : "unrepaired_demage", 
             "dateCreated" : "ad_created", 
             "nrOfPictures" : "nr_of_pictures", 
             "postalCode" : "postal_code", 
             "lastSeen" : "last_seen"}

#use rename method to apply changes across all columns
#inplace to return a new dataframe
autos.rename(names_dict, axis = 1, inplace = True)
#view new names
autos.columns

Index(['Unnamed: 0', 'date_crawled', 'name', 'seller', 'offer_type', 'price',
       'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps',
       'model', 'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_demage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Now that we have more workable names, lets do some data exploration to see what else needs to be tidied up. 

## Exploring the dataset

In [223]:
autos.describe(include = "all")

Unnamed: 0         date_crawled         name  seller offer_type  \
count   50000.000000                50000        50000   50000      50000   
unique           NaN                48213        38754       2          2   
top              NaN  2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot   
freq             NaN                    3           78   49999      49999   
mean    24999.500000                  NaN          NaN     NaN        NaN   
std     14433.901067                  NaN          NaN     NaN        NaN   
min         0.000000                  NaN          NaN     NaN        NaN   
25%     12499.750000                  NaN          NaN     NaN        NaN   
50%     24999.500000                  NaN          NaN     NaN        NaN   
75%     37499.250000                  NaN          NaN     NaN        NaN   
max     49999.000000                  NaN          NaN     NaN        NaN   

        price abtest vehicle_type  registration_year  gearbox  \
count   50000  50000        44905       50000.000000    47320   
unique   2357      2            8                NaN        2   
top        $0   test    limousine                NaN  manuell   
freq     1421  25756        12859                NaN    36993   
mean      NaN    NaN          NaN        2005.073280      NaN   
std       NaN    NaN          NaN         105.712813      NaN   
min       NaN    NaN          NaN        1000.000000      NaN   
25%       NaN    NaN          NaN        1999.000000      NaN   
50%       NaN    NaN          NaN        2003.000000      NaN   
75%       NaN    NaN          NaN        2008.000000      NaN   
max       NaN    NaN          NaN        9999.000000      NaN   

               ...           model   odometer registration_month  fuel_type  \
count          ...           47242      50000       50000.000000      45518   
unique         ...             245         13                NaN          7   
top            ...            golf  150,000km                NaN     benzin   
freq           ...            4024      32424                NaN      30107   
mean           ...             NaN        NaN           5.723360        NaN   
std            ...             NaN        NaN           3.711984        NaN   
min            ...             NaN        NaN           0.000000        NaN   
25%            ...             NaN        NaN           3.000000        NaN   
50%            ...             NaN        NaN           6.000000        NaN   
75%            ...             NaN        NaN           9.000000        NaN   
max            ...             NaN        NaN          12.000000        NaN   

             brand unrepaired_demage           ad_created nr_of_pictures  \
count        50000             40171                50000        50000.0   
unique          40                 2                   76            NaN   
top     volkswagen              nein  2016-04-03 00:00:00            NaN   
freq         10687             35232                 1946            NaN   
mean           NaN               NaN                  NaN            0.0   
std            NaN               NaN                  NaN            0.0   
min            NaN               NaN                  NaN            0.0   
25%            NaN               NaN                  NaN            0.0   
50%            NaN               NaN                  NaN            0.0   
75%            NaN               NaN                  NaN            0.0   
max            NaN               NaN                  NaN            0.0   

         postal_code            last_seen  
count   50000.000000                50000  
unique           NaN                39481  
top              NaN  2016-04-07 06:17:27  
freq             NaN                    8  
mean    50813.627300                  NaN  
std     25779.747957                  NaN  
min      1067.000000                  NaN  
25%     30451.000000                  NaN  
50%     49577.000000                  NaN  
75

### Findings:

- There are a number of text columns where all (or nearly all) of the values are the same: seller, offer_type, and nr_of_pictures. These are likely candidates to be dropped.
- The price and odometer_km variables contain non-numeric characters. Also, the object appears to be stored as an object - probably as a string. These variables can be cleaned then coerced into a numeric dtype. 
- Since the listings were from Germany it seems likely that the values are euros, not dollars. So I will address this by changing price into price_euros, and odometers into odometer_km since Europeans use kilometers instead of miles. 

In [224]:
#removing non-numeric characters and renaming the variables
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
autos.rename({"price" : "price_euro"}, axis = 1, inplace = True)

In [225]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)
autos.rename({"odometer": "odometer_km"}, axis = 1, inplace = True)

In [226]:
autos[["odometer_km", "price_euro"]].head()

odometer_km  price_euro
0       150000        5000
1       150000        8500
2        70000        8990
3        70000        4350
4       150000        1350

In [227]:
#dropping variables
autos.drop(["seller", "offer_type","nr_of_pictures"], axis = 1, inplace = True)

Now, I want to take a closer look into the price and odometer variables to see if I can find any outliers.

In [228]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [229]:
print(str(autos["odometer_km"].unique().shape[0]) + " unique values")
print(autos["odometer_km"].describe())

13 unique values
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [230]:
autos["price_euro"].value_counts().head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_euro, dtype: int64

In [231]:
print(str(autos["price_euro"].unique().shape[0]) + " unique values")
print(autos["price_euro"].describe())

2357 unique values
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_euro, dtype: float64


### Findings:

- The odometer variable has only rounded numbers. This probably isnt an error but a result of eBay's preset options. The website gives sellers a select number of options to choose from when listing their products.
- Most of the cars are used with a fair amount of mileage on them.
- A lot of cars are listed at price 0. This is mostly due to eBay being an auctioning site. Therefore, it seems probable that some of the cars captured in the dataset were intially 0 euros before bidding. But there are some suspicious values: the max is 1e8, which seems a bit too high so I'll do some more digging.

In [232]:
autos["price_euro"].value_counts(normalize = True).sort_index(ascending = False).head(20)

99999999    0.00002
27322222    0.00002
12345678    0.00006
11111111    0.00004
10000000    0.00002
3890000     0.00002
1300000     0.00002
1234566     0.00002
999999      0.00004
999990      0.00002
350000      0.00002
345000      0.00002
299000      0.00002
295000      0.00002
265000      0.00002
259000      0.00002
250000      0.00002
220000      0.00002
198000      0.00002
197000      0.00002
Name: price_euro, dtype: float64

In [233]:
autos["price_euro"].value_counts(normalize = True).sort_index(ascending = True).head(20)

0     0.02842
1     0.00312
2     0.00006
3     0.00002
5     0.00004
8     0.00002
9     0.00002
10    0.00014
11    0.00004
12    0.00006
13    0.00004
14    0.00002
15    0.00004
17    0.00006
18    0.00002
20    0.00008
25    0.00010
29    0.00002
30    0.00014
35    0.00002
Name: price_euro, dtype: float64

It looks like the values jump drastically from 350,000 to 100,000,000 - which likely is due to an error. Therefore, I am going to remove values greater than 350,000 euros and keep everything below it.

In [234]:
autos = autos[autos["price_euro"].between(0,351000)]
autos["price_euro"].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price_euro, dtype: float64

## Exploring the date columns

There are a number of columns with date information: date_crawled, registration_month, registration_year, ad_created, and last_seen. These are a combination of dates that were crawled, and dates with meta-data information attatched to them from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [235]:
autos[["ad_created",
       "date_crawled", 
       "registration_year", 
       "registration_month", "last_seen"]].describe(include = "all")

ad_created         date_crawled  registration_year  \
count                 49986                49986       49986.000000   
unique                   76                48200                NaN   
top     2016-04-03 00:00:00  2016-03-27 22:55:05                NaN   
freq                   1946                    3                NaN   
mean                    NaN                  NaN        2005.075721   
std                     NaN                  NaN         105.727161   
min                     NaN                  NaN        1000.000000   
25%                     NaN                  NaN        1999.000000   
50%                     NaN                  NaN        2003.000000   
75%                     NaN                  NaN        2008.000000   
max                     NaN                  NaN        9999.000000   

        registration_month            last_seen  
count         49986.000000                49986  
unique                 NaN                39472  
top                    NaN  2016-04-07 06:17:27  
freq                   NaN                    8  
mean              5.723723                  NaN  
std               3.711839                  NaN  
min               0.000000                  NaN  
25%               3.000000                  NaN  
50%               6.000000                  NaN  
75%               9.000000                  NaN  
max              12.000000                  NaN

In [236]:
autos[["ad_created",
       "date_crawled", 
       "registration_year", 
       "registration_month", "last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 5 columns):
ad_created            49986 non-null object
date_crawled          49986 non-null object
registration_year     49986 non-null int64
registration_month    49986 non-null int64
last_seen             49986 non-null object
dtypes: int64(2), object(3)
memory usage: 1.7+ MB


In [237]:
autos[["ad_created",
       "date_crawled", 
      "last_seen"]].head()

ad_created         date_crawled            last_seen
0  2016-03-26 00:00:00  2016-03-26 17:47:46  2016-04-06 06:45:54
1  2016-04-04 00:00:00  2016-04-04 13:38:56  2016-04-06 14:45:08
2  2016-03-26 00:00:00  2016-03-26 18:57:24  2016-04-06 20:15:37
3  2016-03-12 00:00:00  2016-03-12 16:58:10  2016-03-15 03:16:28
4  2016-04-01 00:00:00  2016-04-01 14:38:50  2016-04-01 14:38:50

The first ten characters of the ad_created, date_created, and last_seen variable are the year, month, and day; the rest of the characters are meta-data and will be removed.

In [238]:
#looking at the first 10 characters and viewing their relative frequencies
autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21 

In [239]:
autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_values()

2016-01-13    0.000020
2016-02-01    0.000020
2016-01-03    0.000020
2016-01-14    0.000020
2016-02-07    0.000020
2016-01-07    0.000020
2016-02-16    0.000020
2015-08-10    0.000020
2016-02-22    0.000020
2016-02-08    0.000020
2015-11-10    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-29    0.000020
2015-12-30    0.000020
2015-12-05    0.000020
2015-06-11    0.000020
2016-02-11    0.000020
2016-02-17    0.000020
2015-09-09    0.000020
2016-02-02    0.000040
2016-02-20    0.000040
2016-02-09    0.000040
2016-02-05    0.000040
2016-02-24    0.000040
2016-02-26    0.000040
2016-02-14    0.000040
2016-01-10    0.000040
2016-02-18    0.000040
2016-02-12    0.000060
                ...   
2016-03-06    0.015124
2016-03-13    0.016925
2016-03-05    0.023046
2016-03-24    0.029088
2016-03-16    0.030008
2016-03-27    0.030909
2016-03-17    0.031189
2016-03-10    0.031869
2016-03-25    0.031889
2016-03-31    0.031909
2016-03-23    0.032189
2016-03-26    0.032569
2016-03-22 

The ads were created across a variety of dates.

In [240]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()        

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [241]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values()        

2016-04-07    0.001420
2016-04-06    0.003181
2016-03-18    0.013064
2016-04-05    0.013104
2016-03-06    0.013944
2016-03-13    0.015564
2016-03-05    0.025387
2016-03-24    0.029108
2016-03-16    0.029508
2016-03-27    0.031049
2016-03-17    0.031509
2016-03-25    0.031749
2016-03-31    0.031909
2016-03-10    0.032129
2016-03-23    0.032389
2016-03-11    0.032489
2016-03-26    0.032489
2016-03-22    0.032909
2016-03-09    0.033209
2016-03-08    0.033269
2016-03-30    0.033629
2016-04-01    0.033809
2016-03-15    0.033990
2016-03-29    0.034150
2016-03-28    0.034850
2016-03-19    0.034910
2016-04-02    0.035410
2016-03-07    0.035970
2016-04-04    0.036490
2016-03-14    0.036630
2016-03-12    0.036770
2016-03-21    0.037490
2016-03-20    0.037831
2016-04-03    0.038691
Name: date_crawled, dtype: float64

The dates were crawled between March and April of 2016 with regularity.

In [242]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

In [243]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_values()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-18    0.007422
2016-03-08    0.007582
2016-03-13    0.008983
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-14    0.012804
2016-03-19    0.015744
2016-03-15    0.015884
2016-03-27    0.016024
2016-03-16    0.016445
2016-03-26    0.016965
2016-03-23    0.018585
2016-03-25    0.019205
2016-03-24    0.019565
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-28    0.020846
2016-03-22    0.021586
2016-03-29    0.022326
2016-04-01    0.023106
2016-03-12    0.023807
2016-03-31    0.023827
2016-04-04    0.024627
2016-03-30    0.024847
2016-04-02    0.024887
2016-04-03    0.025367
2016-03-17    0.027928
2016-04-05    0.124275
2016-04-07    0.130957
2016-04-06    0.220982
Name: last_seen, dtype: float64

Without more information its hard to extract any insights from the relative frequencies of it being last seen. One thing to notice is the large proportion of viewership during March. But like I said, without more information its hard to understand why that is.

In [244]:
autos["registration_year"].value_counts(normalize = True, dropna = False).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000660
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054055
2002    0.050674
2003    0.054555
2004    0.054755
2005    0.060317
2006    0.054155
2007    0.046093
2008    0.044632
2009    0.041952
2010    0.031949
2011    0.032689
2012    0.026467
2013    0.016125
2014    0.013304
2015    0.007982
2016    0.026327
2017    0.029048
2018    0.009823
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

We have some suspicious values. There are years as low 1000, and as high as 9999. These are definitely mistakes and need to be addressed. Lets take a further look.

In [245]:
autos["registration_year"].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Since the listings are no later than 2016, its impossible to have any cars with a registration year later than 2016. Anything over 2016 will be removed. As for the earliest registration years, its hard to say whether some of them go back as far as 1800. It seems more likely that the listing are within the 1900's. To confirm this I will take a look at the proportion of cars between 1900 and 2016.

In [246]:
(autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.9605889649101749

Less than 4% of the values are before 1900 and after 2016 so I'll remove them.

In [247]:
autos = autos[autos["registration_year"].between(1900,2016)]
print(autos["registration_year"].describe())

print(autos["registration_year"].value_counts(normalize = True).head(20))

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64
2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
2006    0.056377
2001    0.056273
2002    0.052753
1998    0.051087
2007    0.047984
2008    0.046464
2009    0.043673
1997    0.042236
2011    0.034030
2010    0.033260
1996    0.030073
2012    0.027553
2016    0.027408
1995    0.027324
2013    0.016786
Name: registration_year, dtype: float64


### Findings:

- Without more information its hard to extract further insights from the meta-data.
- Most of the cars have a registration year between 1995 and 2016.

Now, lets do some aggregation to make comparisons across categorical variables such as brand.

## Exploring price by brand

In [248]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
citroen           0.013912
smart             0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

### Findings:

- Out of the top 5 most frequent brands in the dataset, 4 of them are German. This makes sense since the listings are from Germany. 
- Volkswagen is the most popular brand listed - over 21% of the listings!


Since many of the brands don't have a significant percentage of listings, we will limit our analysis to brands that are of german origin only - the top 4 brands.

In [249]:
autos["brand"].value_counts(normalize = True).head(4)

volkswagen       0.212117
bmw              0.110026
opel             0.108172
mercedes_benz    0.095364
Name: brand, dtype: float64

In [250]:
brands = autos["brand"].value_counts(normalize = True).head(4) #only german

#creating empty dictionary
brands_dict = {}

#looping
for car in brands.index:
    #selecting cars that match german brands
    selected = autos[autos["brand"] == car]
    #finding mean
    mean = selected["price_euro"].mean()
    #storing car brand as key, and mean as the value
    brands_dict[car] = int(mean)


mean_german_prices = pd.Series(brands_dict).sort_values(ascending = False)
mean_german_prices

mercedes_benz    8485
bmw              8102
volkswagen       5231
opel             2876
dtype: int64

### Findings:

- Mercedes-Benz and BMW are the most expensive.
- Opel is the cheapest of the German brands.
- Volkswagen has the middle-ground price, probably contributing to its popularity.


Now to explore the brands by kilometers.

In [253]:
km_dict = {}

for car in brands.index:
    selected = autos[autos["brand"] == car]
    mean = selected["odometer_km"].mean()
    km_dict[car] = int(mean)

#creating dataframe with brand info
brand_info = pd.DataFrame(mean_german_prices, columns = ["mean_price"])
mean_km = pd.Series(km_dict).sort_values(ascending = False)
brand_info["mean_km"] = mean_km

brand_info

mean_price  mean_km
mercedes_benz        8485   130856
bmw                  8102   132431
volkswagen           5231   128724
opel                 2876   129223

### Findings:

- There is a slight positive trend between price and km. The highest priced German brands also have the highest mileage. This slight trend could be further explored in future analyses.


## Conclusion:

In this analysis we cleaned the dataset by: 
- renaming columns
- removing non-numeric values
- coercing dtypes
- removing outliers

After cleaning the dataset we began to explore the brands by price and kilometers. We also laid the groundwork for further exploration into the meta-data, but we'll leave that for future projects. Goodbye!